geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java

```diff
@@ -203,49 +261,120 @@ public int hstrlen(ByteArrayWrapper field) {
     return new ArrayList<>(hash.keySet());
   }
 
-  public Pair<BigInteger, List<Object>> hscan(Pattern matchPattern, int count, BigInteger cursor) {
-    List<Object> returnList = new ArrayList<Object>();
-    int size = hash.size();
-    BigInteger beforeCursor = new BigInteger("0");
-    int numElements = 0;
-    int i = -1;
-    for (Map.Entry<ByteArrayWrapper, ByteArrayWrapper> entry : hash.entrySet()) {
-      ByteArrayWrapper key = entry.getKey();
-      ByteArrayWrapper value = entry.getValue();
-      i++;
-      if (beforeCursor.compareTo(cursor) < 0) {
-        beforeCursor = beforeCursor.add(new BigInteger("1"));
+  public Pair<BigInteger, List<Object>> hscan(UUID clientID, Pattern matchPattern,
+      int count,
+      BigInteger cursorParameter) {
+
+    int startCursor = cursorParameter.intValue();
+
+    List<ByteArrayWrapper> keysToScan = getSnapShotOfKeySet(clientID);
+
+    List<Object> resultList =
+        getResultList(keysToScan, startCursor, count, matchPattern);
+
+    int indexOfLast = resultList.size() - 1;
+    int numberOfIterationsCompleted = Integer.parseInt(resultList.get(indexOfLast).toString());
+    resultList.remove(indexOfLast);
```

Comment:
See below: rather return a `Pair` so that the intent is clearer.

Reply:
done.  yeah, I really like having to return another another type here, but I don't see a simple way around it. open to suggestions...